internal/state/execution.go

```diff
@@ -414,28 +415,51 @@ func buildLastCommitInfo(block *types.Block, store Store, initialHeight int64) a
 	}
 }
 
-// extendedCommitInfo expects an ExtendedCommitInfo, which includes the
-// original precommit votes along with their vote extensions.
-// Then, it populates the corresponding ABCI protobuf structures.
-func extendedCommitInfo(extCommit types.ExtendedCommit) abci.ExtendedCommitInfo {
-	vs := make([]abci.ExtendedVoteInfo, len(extCommit.ExtendedSignatures))
-	for i, ecs := range extCommit.ExtendedSignatures {
+// buildLastExtendedCommitInfo expects an ExtendedCommitInfo, which includes
+// the original precommit votes along with their vote extensions from the last
+// commit. It populates the corresponding ABCI protobuf structures using the
+// validator set obtained from the store at the height of the last extended
+// commit.
+func buildLastExtendedCommitInfo(ec *types.ExtendedCommit, store Store, initialHeight int64) abci.ExtendedCommitInfo {
+	// The current height is ec.Height+1
+	if ec.Height+1 == initialHeight {
+		// There is no extended commit for the initial height.
+		return abci.ExtendedCommitInfo{}
+	}
+
+	valSet, err := store.LoadValidators(ec.Height)
+	if err != nil {
+		panic(fmt.Errorf("failed to load validator set at height %d: %w", ec.Height, err))
+	}
+
+	vs := make([]abci.ExtendedVoteInfo, ec.Size())
+	for i, ecs := range ec.ExtendedSignatures {
 		var ext []byte
-		if ecs.ForBlock() {
+		if ecs.BlockIDFlag == types.BlockIDFlagCommit {
+			// We only care about vote extensions if a validator has voted to
+			// commit.
 			ext = ecs.VoteExtension
 		}
+		var vpb abci.Validator
+		// The ValidatorAddress field is empty when their vote is absent.
+		if ecs.BlockIDFlag != types.BlockIDFlagAbsent {
+			_, val := valSet.GetByAddress(ecs.ValidatorAddress)
+			if val == nil {
```

Comment:
Actually, this sanity "if" is missing from `buildLastCommitInfo`. Even if this PR is not touching `buildLastCommitInfo`, would you mind adding a similar sanity check there?

Reply:
It's not really needed there. That code relies on the fact that the commit information is in `ValidatorIndex` order and therefore iterates both in the same order, knowing that the indices of the two lists will correspond because the `block`, and therefore `LastCommit` have [just been validated](https://github.com/tendermint/tendermint/blob/b1896ebca6a857c9ec97e8ec086e7ab5d2f0fbf5/internal/state/execution.go#L179).

IMO, this could be done the same way since we [make the extended commit](https://github.com/tendermint/tendermint/blob/b1896ebca6a857c9ec97e8ec086e7ab5d2f0fbf5/internal/consensus/state.go#L1409) just prior to calling this function, which is made using the voteset in [validator index order](https://github.com/tendermint/tendermint/blob/b1896ebca6a857c9ec97e8ec086e7ab5d2f0fbf5/types/vote_set.go#L254). Either way works, but we should add a comment indicating the expectations of `buildLastCommitInfo`, namely that it needs the list to be in validator index order.

Reply:
I've renamed `buildLastExtendedCommitInfo` to just `buildExtendedCommitInfo` (makes more sense given what it does and the parameters it takes) and I've refactored the loop to resemble the one in `buildLastCommitInfo`.

I've left the panic in `buildExtendedCommitInfo` to check if there's a difference between the order of the commit sigs and the validators - this helped pick up a potential issue in `TestMaxProposalBlockSize` where we were faking an extended commit whose sigs were not in validator index order. I'm now constructing this extended commit more realistically.